### PR TITLE
[Feature](inverted index) add parser_mode properties for inverted index parser

### PR DESCRIPTION
**Pull request by airborne12**
# Proposed changes

We add parser mode for inverted index, usage like this:
```
CREATE TABLE `inverted` (
  `FIELD0` text NULL,
  `FIELD1` text NULL,
  `FIELD2` text NULL,
  `FIELD3` text NULL,
  INDEX idx_name1 (`FIELD0`) USING INVERTED PROPERTIES("parser" = "chinese", "parser_mode" = "fine_grained") COMMENT '',
  INDEX idx_name2 (`FIELD1`) USING INVERTED PROPERTIES("parser" = "chinese", "parser_mode" = "coarse_grained") COMMENT ''
) ENGINE=OLAP
);
```

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/inverted_index_writer.cpp [hunk 1]
```diff
@@ -171,6 +171,12 @@ class InvertedIndexColumnWriterImpl : public InvertedIndexColumnWriter {
             auto chinese_analyzer = _CLNEW lucene::analysis::LanguageBasedAnalyzer();
             chinese_analyzer->setLanguage(L"chinese");
             chinese_analyzer->initDict(config::inverted_index_dict_path);
+            auto mode = get_parser_mode_string_from_properties(_index_meta->properties());
+            if (mode == INVERTED_INDEX_PARSER_COARSE_GRANULARITY) {
+                chinese_analyzer->setMode(lucene::analysis::AnalyzerMode::Default);
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index_writer.cpp:176]: Normally check special case in if branches and left default to else branch.

File: be/src/olap/rowset/segment_v2/inverted_index_writer.cpp [hunk 2]
```diff
@@ -215,13 +221,9 @@ class InvertedIndexColumnWriterImpl : public InvertedIndexColumnWriter {
     }
 
     void new_fulltext_field(const char* field_value_data, size_t field_value_size) {
-        if (_parser_type == InvertedIndexParserType::PARSER_ENGLISH) {
+        if (_parser_type == InvertedIndexParserType::PARSER_ENGLISH ||
+            _parser_type == InvertedIndexParserType::PARSER_CHINESE) {
             new_char_token_stream(field_value_data, field_value_size, _field);
-        } else if (_parser_type == InvertedIndexParserType::PARSER_CHINESE) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index_writer.cpp:220]: Is there any risk to merge english and chinese parser?

File: regression-test/suites/inverted_index_p0/test_chinese_analyzer.groovy [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    def indexTblName = "chinese_analyzer_test"
+
+    sql "DROP TABLE IF EXISTS ${indexTblName}"
+    // create 1 replica table
+    sql """
+	CREATE TABLE IF NOT EXISTS ${indexTblName}(
+		`id`int(11)NULL,
+		`c` text NULL,
+		INDEX c_idx(`c`) USING INVERTED PROPERTIES("parser"="chinese","parser_mode"="fine_grained") COMMENT ''
```
Comment (xiaokang) [inline regression-test/suites/inverted_index_p0/test_chinese_analyzer.groovy:36]: Add another column and index with coarse_graind parser_mode and test match an the column

File: be/src/olap/rowset/segment_v2/inverted_index_reader.h [hunk 4]
```diff
@@ -78,76 +78,72 @@ enum class InvertedIndexQueryType {
...
-    uint32_t get_index_id() const { return _index_id; }
+    uint32_t get_index_id() const { return _index_meta.index_id(); }
 
 protected:
     bool _is_match_query(InvertedIndexQueryType query_type);
     friend class InvertedIndexIterator;
     io::FileSystemSPtr _fs;
     std::string _path;
-    uint32_t _index_id;
+    TabletIndex _index_meta;
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/inverted_index_reader.h:107]: It's more concise to move _index_meta from IndexIterator to IndexReader. But the life cycle of IndexIterator, IndexReader, and _index_meta need to be checked more carefully.

Comment (airborne12) [reply be/src/olap/rowset/segment_v2/inverted_index_reader.h:107]: use copy constructor, do not worry about life cycle

File: be/src/olap/rowset/segment_v2/inverted_index_writer.cpp (hunk 1 above)
Comment (airborne12) [reply be/src/olap/rowset/segment_v2/inverted_index_writer.cpp:176]: we want to use All mode as default

File: be/src/olap/rowset/segment_v2/inverted_index_writer.cpp (hunk 2 above)
Comment (airborne12) [reply be/src/olap/rowset/segment_v2/inverted_index_writer.cpp:220]: use char to avoid copy data in both english and chinese analyzer

**Comment by airborne12** [comment]
run buildall

**Review by qidaye — DISMISSED** [review]
LGTM

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Comment by airborne12** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**